File: android/src/main/kotlin/com/example/flutter_thermal_printer/models/PrintableReceipt.kt
package com.example.flutter_thermal_printer.models

import android.util.Log
import com.google.gson.annotations.SerializedName
import java.lang.Integer.min


class PrintableReceipt(
    val address: String?,
    @SerializedName("datetime")
    val datetime: String,
    @SerializedName("delivery_type")
    val deliveryType: String,
    val discount: Double,
    val items: List<CartItem>,
    @SerializedName("order_id")
    val orderId: String,
    @SerializedName("order_total")
    val orderTotal: Double,
    @SerializedName("other_charges")
    val otherCharges: List<OtherCharge>,
    @SerializedName("printer_id")
    val printerId: String,
    @SerializedName("business_name")
    val businessName: String,
    @SerializedName("customer_phone")
    val customerPhone: String

    ) {

    public fun generatePrintableString(qrCodeText: String? = null): String {
        var printableString =
            "[C]<font size='big'>${orderId}</font>\n" +
                    "[C]<b>${datetime}</b>\n" +
                    "[C]<b>${businessName}</b>\n" +
                    "[C]<b>Customer Ph: ${customerPhone}</b>\n" +
                    "[C]--------------------------------\n" +
                    "<b>Items       Qty   Price  Total  </b>\n" +
                    "[C]--------------------------------\n"
        for (orderItem in items) {
            printableString += addOrderItemToPrintableString(orderItem)
        }
        for (charge in otherCharges) {
            printableString += "[R]${charge.name} ${charge.value}\n"
        }
        printableString += "[R]--------------------------------\n"
        printableString += "[R]Rs. ${orderTotal}\n"
        printableString += "[C]--------------------------------\n"
        printableString += "<b>${deliveryType}</b>\n"
        printableString += "[C]--------------------------------\n"
        if (address != null) {
            printableString += "[L]<font size='big'>${address}</font>\n"
        }
        if (qrCodeText != null) {
            printableString += "[C]<qrcode size='20'>${qrCodeText}</qrcode>\n"
        }
        return printableString
    }
    private fun addOrderItemToPrintableString(orderItem: CartItem): String {
        val ITEM_NAME_WIDTH = 12;
        val ITEM_QTY_WIDTH = 6;
        val ITEM_PRICE_WIDTH = 7;
        val ITEM_TOTAL_WIDTH = 7;
        val startIndexed = mutableListOf<Int>(0,0,0,0)
        var printableOrderItemString = ""
        while (true) {
            if (startIndexed[0] == orderItem.name.length &&
                startIndexed[1] == orderItem.quantity.toString().length &&
                startIndexed[2] == orderItem.price.toString().length &&
                startIndexed[3] == orderItem.total.toString().length) break;
            val endIndex1 = min(
                startIndexed[0] + ITEM_NAME_WIDTH - 1,
                orderItem.name.length);
            val name = orderItem.name.substring(startIndexed[0], endIndex1);


            startIndexed[0] = endIndex1;
            printableOrderItemString += name + " " + " ".repeat (ITEM_NAME_WIDTH - name.length - 1)

            val endIndex2 = min(
                startIndexed[1] + ITEM_QTY_WIDTH - 1,
                orderItem.quantity.toString().length);
            val quantity = orderItem.quantity.toString().substring(startIndexed[1], endIndex2);
            startIndexed[1] = endIndex2;
            printableOrderItemString += quantity + " " + " ".repeat (ITEM_QTY_WIDTH - quantity.length - 1)

            val endIndex3 = min(
                startIndexed[2] + ITEM_PRICE_WIDTH - 1,
                orderItem.price.toString().length);
            val price = orderItem.price.toString().substring(startIndexed[2], endIndex3);
            startIndexed[2] = endIndex3;
            printableOrderItemString += price + " " + " ".repeat (ITEM_PRICE_WIDTH - price.length - 1)

            val endIndex4 = min(
                startIndexed[3] + ITEM_TOTAL_WIDTH - 1,
                orderItem.total.toString().length);
            val total = orderItem.total.toString().substring(startIndexed[3], endIndex4);
            startIndexed[3] = endIndex4;
            printableOrderItemString += total + " " + " ".repeat (ITEM_TOTAL_WIDTH - price.length - 1)
            printableOrderItemString += '\n'
        }
        printableOrderItemString += "[C]--------------------------------\n"
        return printableOrderItemString
    }
}

class CartItem(
    val name: String,
    val price: Double,
    val quantity: Int,
    val total: Double
)

data class OtherCharge(
    val name: String,
    val value: Double,
)

File: lib/models/printable_receipt.dart
class PrintableCharges {
  final String name;
  final double value;
  PrintableCharges({
    required this.name,
    required this.value,
  });
  factory PrintableCharges.fromJson(Map<String, dynamic> json) {
    return PrintableCharges(
      name: json['name'] ?? '-',
      value: json['value'].runtimeType == int
          ? (json['value'] as int).paisaToRupee
          : json['value'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'value': value,
    };
  }

  @override
  String toString() => 'PrintableCharges(name: $name, value: $value)';
}

class PrintableOrderItems {
  final String name;
  final double total;
  final int quantity;
  final double price;
  PrintableOrderItems({
    required this.name,
    required this.total,
    required this.quantity,
    required this.price,
  });

  factory PrintableOrderItems.fromJson(Map<String, dynamic> json) {
    return PrintableOrderItems(
      name: json['name'] ?? '-',
      total: json['total'] != null
          ? json['total'].runtimeType == int
              ? (json['total'] as int).paisaToRupee
              : json['total']
          : 0,
      quantity: json['quantity'] ?? 0,
      price: json['price'] != null
          ? json['price'].runtimeType == int
              ? (json['price'] as int).paisaToRupee
              : json['price']
          : 0,
    );
  }

  @override
  String toString() {
    return 'PrintableOrderItems(name: $name, total: $total, quantity: $quantity, price: $price)';
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'total': total,
      'quantity': quantity,
      'price': price,
    };
  }
}

class PrintableReceipt {
  final String dateTime;
  final String address;
  final String deliveryType;
  final List<PrintableOrderItems> items;
  final List<PrintableCharges> otherCharges;
  final double discount;
  final double orderTotal;
  final String printerId;
  final String businessName;
  final String orderId;
  final String customerPhone;
  PrintableReceipt({
    required this.dateTime,
    required this.customerPhone,
    required this.address,
    required this.deliveryType,
    required this.items,
    required this.otherCharges,
    required this.discount,
    required this.orderTotal,
    required this.printerId,
    required this.businessName,
    required this.orderId,
  });
  factory PrintableReceipt.fromJson(Map<String, dynamic> json) {
    return PrintableReceipt(
      dateTime: json['datetime'] ?? '-',
      address: json['address'] ?? '-',
      deliveryType: json['delivery_type'] ?? '-',
      businessName: json['business_name'] ?? '-',
      items: json['items'] == null
          ? []
          : (json['items'] as List)
              .map((e) => PrintableOrderItems.fromJson(e))
              .toList(),
      otherCharges: json['other_charges'] == null
          ? []
          : (json['other_charges'] as List)
              .map((e) => PrintableCharges.fromJson(e))
              .toList(),
      discount: json['discount'] != null
          ? json['discount'].runtimeType == int
              ? (json['discount'] as int).paisaToRupee
              : json['discount']
          : 0,
      orderTotal: json['order_total'] != null
          ? json['order_total'].runtimeType == int
              ? (json['order_total'] as int).paisaToRupee
              : json['order_total']
          : 0,
      printerId: json['printer_id'] ?? '',
      orderId: json['order_id'] ?? '-',
      customerPhone: json['customer_phone'] ?? '-',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'date_time': dateTime,
      'address': address,
      'delivery_type': deliveryType,
      'items': items.map((x) => x.toJson()).toList(),
      'other_charges': otherCharges.map((x) => x.toJson()).toList(),
      'discount': discount,
      'order_total': orderTotal,
      'printer_id': printerId,
      'business_name': businessName,
      'order_id': orderId,
      'customer_phone': customerPhone,
    };
  }
}

extension ConvertPaisaToRupee on num? {
  double get paisaToRupee => this == null ? 0 : this! / 100;
}
